compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangListConstructorExpr.java

```diff
@@ -35,7 +35,7 @@
  * @see BLangJSONArrayLiteral
  * @since 1.0
  */
-public class BLangListConstructorExpr extends BLangExpression implements ListConstructorExprNode {
+public class BLangListConstructorExpr extends BLangAccessExpression implements ListConstructorExprNode {
```

Comment:
I'm not sure if we should be marking these as `BLangAccessExpression`s. These aren't access expressions, right? They are the expressions on which the access is done?

Shouldn't we fix the visits for field/member access instead? To consider any `BLangExpression` rather than expecting a `BLangVariableReference` (https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L2465)?

Reply:
I have a mixed opinion on this. But, if we have any better way as @MaryamZi mentioned, then let's explore that and see.

Reply:
Is it ok to introduce another abstraction with `lhsVar`, `compoundAssignmentLhsVar`, etc..., and expect that? Or is it possible to get rid of analyzing them all together, given that we need to allow all the BLangExpression? Or do we have to analyze case by case (i.e BLangVariableReferences, Structural Constructs, etc...)?

Reply:
I believe we still need to preserve information regarding whether or not certain expressions are on the LHS to support/restrict certain scenarios.

```ballerina
type Foo record {|
    int i?;
|};

public function main() {
    Foo f = {};

    f.i = 1; // field access for an optional field on the LHS - valid
    int? i = f.i; // field access for an optional field on the RHS - invalid

    int? j = (<Foo> {})?.i; // valid

    (<Foo> {}).i = 2; // invalid, LHS is not a variable-reference-lvexpr, field-access-lvexpr, nor a member-access-lvexpr
}
```

The spec has this concept of lvalues (https://ballerina.io/spec/lang/master/#section_7.14.1) and an `lvexpr` is defined as 

```
lvexpr :=
   variable-reference-lvexpr
   | field-access-lvexpr
   | member-access-lvexpr
```

Thinking out loud, `lhsVar`, `compoundAssignmentLhsVar`, etc. may thus only be relevant for `variable-reference`, `field-access`, and `member-access`. 

I think we should look into the possibility of basing the abstraction on this.

Reply:
> Thinking out loud, lhsVar, compoundAssignmentLhsVar, etc. may thus only be relevant for variable-reference, field-access, and member-access.

@MaryamZi, I agree, but with the current implementation, `BLangAccessExpression` extends `BLangVariableReference` and both `BLangFieldBasedAccess`, `BLangIndexBasedAccess` extends `BLangAccessExpression`. Since, information like `lhsVar` & `compoundAssignmentLhsVar` belongs to `BLangVariableReference`, there's already a such abstraction. However, it's not only limited to `lvexpr`, but inherits by each and every expression which descendant from `BLangVariableReference`. Is that acceptable?